snapshot/snapshot.go

```diff
@@ -757,23 +832,67 @@ func (o *snapshotter) mounts(ctx context.Context, s storage.Snapshot, checkKey s
 		}, nil
 	}
 
-	parentPaths := make([]string, len(s.ParentIDs))
-	for i := range s.ParentIDs {
-		parentPaths[i] = o.upperPath(s.ParentIDs[i])
+	parentPaths, err := o.getParentPaths(ctx, s)
+	if err != nil {
+		return nil, err
 	}
 
 	options = append(options, fmt.Sprintf("lowerdir=%s", strings.Join(parentPaths, ":")))
 	if o.userxattr {
 		options = append(options, "userxattr")
 	}
+
 	return []mount.Mount{
 		{
 			Type:    "overlay",
 			Source:  "overlay",
 			Options: options,
 		},
 	}, nil
+}
+
+func (o *snapshotter) getParentPaths(ctx context.Context, s storage.Snapshot) ([]string, error) {
+	parentPaths := make([]string, len(s.ParentIDs))
+
+	for i := range s.ParentIDs {
+		id := s.ParentIDs[i]
+		if _, ok := o.idmapped[s.ID]; ok {
+			id = fmt.Sprintf("%s_%s", s.ParentIDs[i], s.ID)
+		}
+		parentPaths[i] = o.upperPath(id)
+	}
+
+	return parentPaths, nil
+}
+
+func (o *snapshotter) createIDMapMounts(ctx context.Context, s storage.Snapshot, idmap idtools.IDMap) error {
+	log.G(ctx).Infof("mapping ids")
```

Comment:
Formatting is not required, but also is this logging too verbose? Perhaps consider removing or reducing to trace.